packages/trader/src/App/Components/Elements/market-countdown-timer.jsx

```diff
@@ -0,0 +1,156 @@
+        }
+    }
+};
+// eslint-disable-next-line consistent-return
+const whenMarketOpens = async (days_offset, target_symbol) => {
+    // days_offset is 0 for today, 1 for tomorrow, etc.
+    if (days_offset > days_to_check_before_exit) return undefined;
+    let when_market_opens;
+    const date_target = new Date();
+    date_target.setDate(date_target.getDate() + days_offset);
+    const date = moment(date_target).format('YYYY-MM-DD');
+    const api_res = await getTradingTimes(date);
+    if (!api_res.api_initial_load_error) {
+        const { times } = getSymbol(target_symbol, api_res.trading_times);
+        const { open, close } = times;
+        const is_closed_all_day = open?.length === 1 && open[0] === '--' && close[0] === '--';
+        if (is_closed_all_day) {
+            // check tomorrow trading times
+            when_market_opens = await whenMarketOpens(days_offset + 1, target_symbol);
+            return when_market_opens;
+        }
+        const date_str = date_target.toISOString().substring(0, 11);
+        const getUTCDate = hour => new Date(`${date_str}${hour}Z`);
+        for (let i = 0; i < open?.length; i++) {
+            const diff = +getUTCDate(open[i]) - +new Date();
+            if (diff > 0) {
+                when_market_opens = +getUTCDate(open[i]);
+                return when_market_opens;
+            }
+        }
+        when_market_opens = await whenMarketOpens(days_offset + 1, target_symbol);
+        return when_market_opens;
+    }
+};
+
+const calculateTimeLeft = when_market_opens => {
+    const difference = when_market_opens - +new Date();
+    let timeLeft = {};
+
+    if (difference > 0) {
+        timeLeft = {
+            days: Math.floor(difference / (1000 * 60 * 60 * 24)),
+            hours: Math.floor((difference / (1000 * 60 * 60)) % 24),
+            minutes: Math.floor((difference / 1000 / 60) % 60),
+            seconds: Math.floor((difference / 1000) % 60),
+        };
+    }
+
+    return timeLeft;
+};
+
+const padWithZero = number => {
+    if (number < 10) {
+        return `0${number}`;
+    }
+    return number;
+};
+
+const MarketCountdownTimer = ({ is_main_page, symbol }) => {
+    const [when_market_opens, setWhenMarketOpens] = React.useState('');
+    const [time_left, setTimeLeft] = React.useState(calculateTimeLeft(when_market_opens));
+
+    React.useEffect(() => {
+        let is_subscribed = true;
+        async function fetchTradingTimes() {
+            const response = await whenMarketOpens(0, symbol);
+            if (is_subscribed) setWhenMarketOpens(response);
+        }
+        fetchTradingTimes();
+
+        return () => (is_subscribed = false);
+    }, []);
+
+    React.useEffect(() => {
+        const timer = setTimeout(() => {
+            setTimeLeft(calculateTimeLeft(when_market_opens));
+        }, 1000);
+
+        return () => {
+            if (timer) {
+                clearTimeout(timer);
+            }
+        };
+    }, [time_left]);
```

Comment:
This relies on `when_market_opens` we should include it in the useEffect dependencies.

Reply:
Yes but it will stop timer functionality.

Reply:
Shouldn't this case be handled by moving `timer` out of the effect? OR initialize `when_market_open` with `calculateTimeLeft('')`?

Reply:
Surprisingly adding `when_market_open` as a dependency worked! Previously didn't, BTW this component has seen many changes since its birth.